src/vm/tailcallhelp.cpp

```diff
@@ -0,0 +1,931 @@
+            pSig, cbSig,
+            &emptyCtx,
+            &sl);
+
+#ifdef _DEBUG
+    StackSString ilStub;
+    sl.LogILStub(CORJIT_FLAGS(), &ilStub);
+    StackScratchBuffer ssb;
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: CallTarget IL created\n"));
+    LOG((LF_STUBS, LL_INFO1000, "%s\n\n", ilStub.GetUTF8(ssb)));
+#endif
+
+    return pCallTargetMD;
+}
+
+void TailCallHelp::CreateCallTargetStubSig(const TailCallInfo& info, SigBuilder* sig)
+{
+    sig->AppendByte(IMAGE_CEE_CS_CALLCONV_DEFAULT);
+
+    // Arg buffer, return value pointer, and pointer to "tail call aware return address" field.
+    sig->AppendData(3);
+
+    // Returns void
+    sig->AppendElementType(ELEMENT_TYPE_VOID);
+
+    // Arg buffer
+    sig->AppendElementType(ELEMENT_TYPE_I);
+
+    // Return value
+    sig->AppendElementType(ELEMENT_TYPE_I);
+
+    // Pointer to tail call aware return address
+    sig->AppendElementType(ELEMENT_TYPE_I);
+
+#ifdef _DEBUG
+    DWORD cbSig;
+    PCCOR_SIGNATURE pSig = (PCCOR_SIGNATURE)sig->GetSignature(&cbSig);
+    SigTypeContext emptyContext;
+    MetaSig outMsig(pSig, cbSig, info.CallSiteSig->GetModule(), &emptyContext);
+    SigFormat outSig(outMsig, NULL);
+    LOG((LF_STUBS, LL_INFO1000, "TAILCALLHELP: CallTarget sig: %s\n", outSig.GetCString()));
+#endif // _DEBUG
+}
+
+void TailCallHelp::EmitLoadTyHnd(ILCodeStream* stream, TypeHandle tyHnd)
+{
+    CorElementType ty = tyHnd.GetSignatureCorElementType();
+    if (tyHnd.IsByRef())
+    {
+        // Note: we can use an "untracked" ldind.i here even with byrefs because
+        // we are loading between two tracked positions.
+        stream->EmitLDIND_I();
+    }
+    else
+    {
+        int token = stream->GetToken(tyHnd);
+        stream->EmitLDOBJ(token);
+    }
+}
+
+void TailCallHelp::EmitStoreTyHnd(ILCodeStream* stream, TypeHandle tyHnd)
+{
+    CorElementType ty = tyHnd.GetSignatureCorElementType();
+    if (tyHnd.IsByRef())
+    {
+        // Note: we can use an "untracked" stind.i here even with byrefs because
+        // we are storing between two tracked positions.
+        stream->EmitSTIND_I();
+    }
+    else
+    {
+        int token = stream->GetToken(tyHnd);
+        stream->EmitSTOBJ(token);
+    }
+}
+
+void TailCallHelp::AppendTypeHandle(SigBuilder& builder, TypeHandle th)
+{
+    if (th.IsByRef())
+    {
+        builder.AppendElementType(ELEMENT_TYPE_BYREF);
+        th = th.AsTypeDesc()->GetBaseTypeParam();
+    }
+
+    CorElementType ty = th.GetSignatureCorElementType();
+    if (CorTypeInfo::IsPrimitiveType(ty) ||
+        ty == ELEMENT_TYPE_OBJECT || ty == ELEMENT_TYPE_STRING)
+    {
+        builder.AppendElementType(ty);
+        return;
+    }
+
+    _ASSERTE(ty == ELEMENT_TYPE_VALUETYPE || ty == ELEMENT_TYPE_CLASS);
+    builder.AppendElementType(ELEMENT_TYPE_INTERNAL);
+    builder.AppendPointer(th.AsPtr());
+}
+
+PCCOR_SIGNATURE TailCallHelp::AllocateSignature(LoaderAllocator* alloc, SigBuilder& sig, DWORD* sigLen)
```

Comment:
A nit - can you please use `pLoaderAllocator` instead of `alloc` name here and in the method below?